packages/astro/src/core/preview/vite-plugin-astro-preview.ts

```diff
@@ -53,29 +52,49 @@ export function vitePluginAstroPreview(settings: AstroSettings): Plugin {
 			});
 
 			return () => {
-				const fourOhFourMiddleware: Connect.NextHandleFunction = (req, res) => {
-					const errorPagePath = fileURLToPath(outDir + '/404.html');
-					if (fs.existsSync(errorPagePath)) {
-						res.statusCode = 404;
-						res.setHeader('Content-Type', 'text/html;charset=utf-8');
-						res.end(fs.readFileSync(errorPagePath));
-					} else {
-						const pathname = stripBase(req.url!, base);
-						res.statusCode = 404;
-						res.end(notFoundTemplate(pathname, 'Not Found'));
-					}
-				};
+				// NOTE: the `base` is stripped from `req.url` for post middlewares
 
-				// Vite 5 has its own 404 middleware, we replace it with ours instead.
-				if (IS_VITE_5) {
-					for (const middleware of server.middlewares.stack) {
-						// This hardcoded name will not break between Vite versions
-						if ((middleware.handle as Connect.HandleFunction).name === 'vite404Middleware') {
-							middleware.handle = fourOhFourMiddleware;
+				server.middlewares.use((req, res, next) => {
+					const pathname = cleanUrl(req.url!);
+
+					// Vite doesn't handle /foo/ if /foo.html exists, we handle it anyways
+					if (pathname.endsWith('/')) {
+						const pathnameWithoutSlash = pathname.slice(0, -1);
+						const htmlPath = fileURLToPath(outDir + pathnameWithoutSlash + '.html');
+						if (fs.existsSync(htmlPath)) {
+							req.url = pathnameWithoutSlash + '.html';
+							return next();
 						}
```

Comment:
The comment above should clarify what this does. But to further explain why we're assigning to `req.url`, this is for a later Vite middleware to load the files. We only need to update the `req.url` so that middleware can detect a HTML file.